### ISSUE TITLE
Document how to write a grammar or theme.

### ISSUE THREAD

**Issue by Armitxes** [Microsoft/vscode#2058]
I would like to suggest that we add some default syntax/theme scopes to the docs for basic stuff (classes, methods/functions, operators, etc.) which all themes shall support.
- For example Sublime has: https://gist.github.com/alehandrof/5361546

The issue is that every theme and extension maker supports it's own scopes - but with users changing the themes and/or extensions, the highlighters break / get unsupported.

Even with the last VS Update, I got reports that the highlighting is no longer working while the extension itself hadn't any changes published. Later I noticed that long-time supported scopes such as "entity.name.class" suddenly weren't supported anymore by themes such as the high-contrast one.

Also, scopes such as "support.function" work on some themes like MonoKai or Monokai Dimmed but fail to work with others, again like the high-contrast one.

This creates misunderstandings when working in a team with people who use different themes and makes compatibility harder.

**Comment by aeschli** [comment]
Not an easy request as we're not really in control of the tokenizers that generate all the scopes.
We are using standard TextMate grammars created and maintained by the community. Using TextMate grammars has great benefits. There is a large number of grammars around, for almost every languages, and a driving community making improvements. However, when diving deeper into this, on realizes that writing a grammar or a theme is difficult. What scopes there are and what scopes to use where is not well defined. As a result of the lack of specification also existing grammars differ widely. I found that one has to learn by looking at existing grammars and themes, and by lots of testing.

Resources I found helpful and use as a reference:
 - http://manual.macromates.com/en/language_grammars
 - http://www.apeth.com/nonblog/stories/textmatebundle.html
 - https://tmtheme-editor.herokuapp.com/#!/editor/theme/Monokai

Besides that we also have some VSCode specific problems and known bugs:
- we have coloring rules for non-default scopes that interfere with standard scopes. We are working on removing these rules
- while TextMate scopes are actually hierarchical, internally we match against a flattened representation. This leads to bugs when parent scopes use scope names that overlap with scopes on the token level. This again is something we are trying to fix.

Given that we are working on fixing our own bugs, the best advise for grammar and theme authors is to follow TextMate guidelines and study existing grammars and scopes. Test against other TextMate compatible editors as to avoid any VSCode specific workarounds.
